File: api-server/app/routers/categories.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas.category import (
    CategoryCreate,
    CategoryListResponse,
    CategoryResponse,
    CategoryTreeResponse,
    CategoryUpdate,
)
from app.services import category_service

router = APIRouter(prefix="/api/v1/categories", tags=["Categories"])


@router.post("", response_model=CategoryResponse, status_code=201)
async def create_category(data: CategoryCreate, db: AsyncSession = Depends(get_db)):
    return await category_service.create_category(db, data)


@router.get("")
async def list_categories(
    tree: bool = Query(False, description="Return nested tree structure"),
    db: AsyncSession = Depends(get_db),
):
    categories = await category_service.list_categories(db, tree=tree)
    if tree:
        return [CategoryTreeResponse.model_validate(c) for c in categories]
    total = await category_service.get_category_count(db)
    return CategoryListResponse(
        items=[CategoryResponse.model_validate(c) for c in categories],
        total=total,
    )


@router.get("/{category_id}", response_model=CategoryResponse)
async def get_category(category_id: uuid.UUID, db: AsyncSession = Depends(get_db)):
    return await category_service.get_category(db, category_id)


@router.put("/{category_id}", response_model=CategoryResponse)
async def update_category(
    category_id: uuid.UUID, data: CategoryUpdate, db: AsyncSession = Depends(get_db)
):
    return await category_service.update_category(db, category_id, data)


@router.delete("/{category_id}", status_code=204)
async def delete_category(category_id: uuid.UUID, db: AsyncSession = Depends(get_db)):
    await category_service.delete_category(db, category_id)


@router.get("/{category_id}/children", response_model=list[CategoryResponse])
async def get_children(category_id: uuid.UUID, db: AsyncSession = Depends(get_db)):
    return await category_service.get_children(db, category_id)


@router.get("/{category_id}/ancestors", response_model=list[CategoryResponse])
async def get_ancestors(category_id: uuid.UUID, db: AsyncSession = Depends(get_db)):
    return await category_service.get_ancestors(db, category_id)

File: web-app/src/api/categories.ts
import api from './client';
import type { Category, AttributeDefinition } from '../types';

export async function getCategoryTree(): Promise<Category[]> {
  const { data } = await api.get('/categories?tree=true');
  return data;
}

export async function getCategory(id: string): Promise<Category> {
  const { data } = await api.get(`/categories/${id}`);
  return data;
}

export async function createCategory(payload: {
  name: string;
  description?: string;
  labels?: Record<string, string>;
  parent_id?: string;
}): Promise<Category> {
  const { data } = await api.post('/categories', payload);
  return data;
}

export async function updateCategory(
  id: string,
  payload: Partial<{ name: string; description: string; labels: Record<string, string>; is_active: boolean }>,
): Promise<Category> {
  const { data } = await api.put(`/categories/${id}`, payload);
  return data;
}

export async function deleteCategory(id: string): Promise<void> {
  await api.delete(`/categories/${id}`);
}

export async function getCategoryAttributes(id: string): Promise<AttributeDefinition[]> {
  const { data } = await api.get(`/categories/${id}/attributes`);
  return data;
}

export async function createAttribute(
  categoryId: string,
  payload: { name: string; attribute_type: string; is_required?: boolean; options?: string[]; labels?: Record<string, string> },
): Promise<AttributeDefinition> {
  const { data } = await api.post(`/categories/${categoryId}/attributes`, payload);
  return data;
}
